File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.cs
using System;
using System.Collections.Generic;

public class Solution 
{
    public int[] solution(string[] park, string[] routes) 
    {
        int[] start = new int[2];
        Dictionary<string, int[]> dirDic = new Dictionary<string, int[]>();
        
        dirDic.Add("N", new int[]{-1,0});
        dirDic.Add("S", new int[]{1,0});
        dirDic.Add("W", new int[]{0,-1});
        dirDic.Add("E", new int[]{0,1});

        
        for (int i =0; i<park.Length; i++)
        {
            for (int j = 0; j < park[i].Length ; j++)
            {
                if(park[i][j] == 'S'){
                    start[0]= i;
                    start[1]= j;
                }
            }    
        }

        
        foreach(string route in routes)
        {
            int[] dir = dirDic[route.Substring(0,1)];
            int move = int.Parse(route.Substring(2,1));
            bool isMove = true;
            int[] temp = new int[2];
            Array.Copy(start,temp,temp.Length);
            
            for(int i =0; i< move; i++)
            {
                int nextY = temp[0] + dir[0];
                if ( nextY < 0 || nextY >= park.Length)
                {
                    isMove = false;
                    // 0보다 크고 park.Length 보단 작아야함
                }
                
                int nextX = temp[1] + dir[1];
                if (nextX < 0 || nextX >= park[0].Length)
                {
                    isMove = false;
                }
                
                if (isMove)
                {
                    if (park[nextY][nextX] == 'X'){
                        isMove = false;
                    }
                }
                
                temp[0]= nextY;
                temp[1]= nextX;
                
            }
            
            if (isMove)
            {
                start = temp;       
            }

            //주어진방향으로 이동하는 도중에 장애물이 있는지
            //끝까지 이동할때 공원을 벗어나는지 확인
        }
        

        return start;
    }
}

File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer{0,0};
    int H = park.size();
    int W = park[0].length();
    
    for (int i =0; i<H ; i++){
        int W_index = park[i].find("S");
        
        if( W_index != string::npos)
        {
            answer[0] = i;
            answer[1] = W_index;
            break;
        }
    }
    
    for(string route : routes){
        string direction = route.substr(0,1);
        int move = stoi(route.substr(2,1));
        bool check_index = true;
        bool check_wall = true;

        switch(direction[0]){
            case 'N' :
                if(answer[0] - move < 0)
                    check_index = false;
                break;
            case 'S' :
                if(answer[0] + move >= H)
                    check_index = false;
                break;
            case 'W' :
                if(answer[1] - move < 0)
                    check_index = false;
                break;
            case 'E' :
                if(answer[1] + move >= W)
                    check_index = false;
                break;
        }
    
        if (!check_index) continue;
        
        switch(direction[0]){
        case 'N' :
            for (int i =1; i <= move; i++){
                if(park[answer[0]-i][answer[1]] == 'X'){
                    check_wall= false;
                    break;
                }
                    
            }
            break;
        case 'S':
            for(int i =1; i<= move; i++){
                if(park[answer[0]+i][answer[1]] == 'X'){
                    check_wall= false;
                    break;
                }
                    
            }
            break;
        case 'W':
            for(int i =1; i<= move; i++){
                if(park[answer[0]][answer[1]-i] == 'X'){
                    check_wall= false;
                    break;
                }
            }
            break;
        case 'E':
            for(int i =1; i<= move; i++){
                if(park[answer[0]][answer[1]+i] == 'X'){
                    check_wall= false;
                    break;
                }
            }
            break;
    }
        
    if (!check_wall) continue;
        
    switch(direction[0]){
        case 'N' :
            answer[0] -= move;
            break;
        case 'S':
            answer[0] += move;
            break;
        case 'W':
            answer[1] -= move;
            break;
        case 'E':
            answer[1] += move;
            break;
    }

        
    }
    
    return answer;
}



    
